### PR TITLE
mount and swap improvements

### PR DESCRIPTION
**Pull request by poettering**
Primarily a fix for #10874 but many other smaller improvements too.

### CODE REVIEW

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging ca7d7db84f52eaf2cc9ba1863c04d4254b026ae2 into 50ae773f85216875033ae47362bb906cf5f5b2d8 - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-516391be8061dc8be4d3b0d083e7569c28ddfda4)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/core/mount.c [hunk 1]
```diff
@@ -1598,20 +1608,12 @@ static int mount_setup_unit(
...
-                 * sysadmin having called mount(8) directly. */
-                r = unit_new_for_name(m, sizeof(Mount), e, &u);
-                if (r < 0)
-                        goto fail;
-
-                r = mount_setup_new_unit(u, what, where, options, fstype, &flags);
-                if (r < 0)
-                        unit_free(u);
-        } else
+        if (!u)
```
Comment (keszybz) [inline src/core/mount.c]: Maybe it'd be even nicer to avoid the reverted condition here: `if (u) ...`.

File: src/core/mount.c [hunk 2]
```diff
@@ -1483,6 +1483,11 @@ static int mount_setup_new_unit(
         if (r < 0)
                 return r;
 
+        /* This unit was generated because /proc/self/mountinfo reported it. Remember this, so that by the time we load the */
```
Comment (keszybz) [inline src/core/mount.c]: ... please don't leave the reader hanging here!

File: src/core/mount.c [hunk 3]
```diff
@@ -1529,35 +1527,23 @@ static int mount_setup_existing_unit(
...
 
-        if (load_extras)
-                return mount_add_extras(MOUNT(u));
+        if (flags->just_changed) {
+                /* If things changed, then make sure that all deps are regenerated. Let's
+                 * first remove all automatic deps, and then add in the new ones. */
+
+                unit_remove_dependencies(u, UNIT_DEPENDENCY_MOUNTINFO_IMPLICIT);
+
+                r = mount_add_extras(MOUNT(u));
```
Comment (keszybz) [inline src/core/mount.c:1551]: Nice!

**Comment by poettering** [comment]
Force pushed a new version. Only made the requested changes. Given how trivial the changes were (one comment, and one if branch reordering) taking liberty to set green label.

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging 06721f39f6adcfd8319ffb3d79ece87b8ef38506 into 1478aa4eb58d5e7406ab608c1cfe3e69543ad47e - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-78da844716d7868e10967ca11d50d04850e963c1)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**